gadopt/utility.py

```diff
@@ -302,6 +307,113 @@ def get_functionspace(mesh, h_family, h_degree, v_family=None, v_degree=None,
     return constructor(mesh, elt, **kwargs)
 
 
+class DiffusiveSmoothingSolver:
+    def __init__(
+            self,
+            function_space: fd.FunctionSpace,
+            wavelength: Number,
+            bcs: Optional[dict[int, dict[str, int | float]]] = None,
+            K: fd.Function | Number = 1,
+            quad_degree: int = 6,
+            solver_parameters: Optional[dict[str, str | float]] = None,
+            **kwargs):
+        """A class to perform diffusive smoothing using a diffusion solver.
+
+        This class provides functionality to solve a diffusion equation for smoothing
+        a Scalar Function.
+
+        Args:
+            function_space (firedrake.FunctionSpace): The function space to solve the diffusion equation.
+            wavelength (Number): The wavelength or time duration for diffusion.
+            bcs (Optional[dict[int, dict[str, int | float]]]): Boundary conditions to impose on the solution.
+            K (Union[firedrake.Function, Number], optional): Diffusion tensor. Defaults to 1.
+            quad_degree (int, optional): Quadrature degree for integrations. Defaults to 6.
+            solver_parameters (Optional[dict[str, Union[str, float]]], optional): Solver parameters for the solver. Defaults to None.
+            **kwargs: Additional keyword arguments to pass to the solver.
+
+        Attributes:
+            function_space (firedrake.FunctionSpace): The function space to solve the diffusion equation.
+            rhs (firedrake.Function): The right-hand side for the diffusion equation.
+            u (firedrake.Function): The solution of the diffusion problem.
+            mesh (firedrake.Mesh): The mesh to solve the problem on.
+            delta_t (Number): The wavelength or time duration for diffusion.
+            bcs (Optional[dict[int, dict[str, int | float]]]): Boundary conditions to impose on the solution.
+            quad_degree (int): Quadrature degree for integrations.
+            solver_parameters (Optional[dict[str, Union[str, float]]]): Solver parameters for the solver.
+            K (Union[firedrake.Function, Number]): Diffusion tensor.
+            solver_kwargs (dict): Additional keyword arguments to pass to the solver.
+            _solver_is_setup (bool): Flag indicating if the solver is set up.
+
+        """
```

Comment:
This docstring should go under the class itself (not in `__init__`), and the `Args` section doesn't need to replicate the types: you've already annotated them in the function definition itself!

If you don't expect the attributes to be directly accessed (which I don't think you do), I don't think they need to be in the docstring like this either.